### PR TITLE
Don't init dist when world_size is 1

### PR DESCRIPTION
**Pull request by jbloxham**
The distributed process group doesn't necessarily need to be initialized when we're doing single-process training, and this can help avoid crashes in certain environments. Note that the DDP initialization code already is a no-op when the distributed process group is not initialized.

Tested by running models with various world sizes (and no explicit world size), and using a print statement to verify whether the model was the original model class or a DDP wrapper.

### CODE REVIEW

File: composer/utils/dist.py [hunk 1]
```diff
@@ -74,7 +74,6 @@ def get_local_rank() -> int:
         int: The local world size
     """
     local_rank = _get_distributed_config_var(env_var="LOCAL_RANK", human_name="local rank", default=0)
-    assert local_rank == get_global_rank() % get_local_world_size(), "invariant violation"
```
Comment (jbloxham) [inline composer/utils/dist.py:77]: This change is unrelated to the main purpose of this PR, but I didn't feel like creating a new PR for it. Essentially, we shouldn't be assuming that all nodes have the same amount of processes.

Comment (ravi-mosaicml) [reply composer/utils/dist.py:77]: If we are removing this assumption, we should also add in a `NODE_RANK` env variable, and update `get_node_rank` to use that.

File: composer/utils/dist.py [hunk 2]
```diff
@@ -235,23 +234,18 @@ def initialize_dist(backend: str, timeout: datetime.timedelta):
         return
 
     if dist.is_initialized():
-        if not dist.get_backend() == backend.lower():
+        if dist.get_backend() != backend.lower() and get_world_size() > 1:
             warnings.warn(f"The requested backend ({backend}) differs from the backend "
```
Comment (ravi-mosaicml) [inline composer/utils/dist.py:238]: Can you convert this back to a RuntimeError? Another PR made it a warning :disappointed:

**Review by ravi-mosaicml — APPROVED** [review]
LGTM

File: composer/utils/dist.py (hunk 1 above)
Comment (jbloxham) [reply composer/utils/dist.py:77]: I'm game.

File: composer/utils/dist.py [hunk 3]
```diff
@@ -235,23 +236,18 @@ def initialize_dist(backend: str, timeout: datetime.timedelta):
         return
 
     if dist.is_initialized():
-        if not dist.get_backend() == backend.lower():
-            warnings.warn(f"The requested backend ({backend}) differs from the backend "
-                          f"of the current process group ({dist.get_backend()})."
-                          "If you wish to change backends, please restart the python process.")
+        if dist.get_backend() != backend.lower() and get_world_size() > 1:
```
Comment (ravi-mosaicml) [inline composer/utils/dist.py:239]: This should apply whenever dist is initialized

```suggestion
        if dist.get_backend() != backend.lower():
```

**Pull request merged**